### ISSUE TITLE
From where to download a zip or tar archive with psql prebuilt binaries?

### ISSUE THREAD

**Issue by acecilia** [PostgresApp/PostgresApp#636]
Hello 👋 

As part of a CI automated job, I need to run `psql`. But first I need to install it. I would like to avoid building the command line tool with brew, so I thought the best idea would be to download a prebuilt binary. 

I see the postgress releases have a `dmg` file attached that contains the prebuilt binary in `Postgres.app/Contents/Versions/13/bin/psql`. The problem is that I cant figure out a way of extracting anything from inside the `dmg` file using the command line. Because the dmg requires mounting, its not a good solution because in case of script failure the image will stay mounted in the machine.

Is there any other mean of distribution that I can use that uses other type of compression (`zip`, `tar`...) and contains the `psql` binary?

Thanks

**Comment by jakob** [comment]
We only provide the DMG files for download. It's a lot of effort to produce them for every release, since the process is only partially automated. Adding additional formats would be even more effort.

My recommendation would be to manually extract psql from Postgres.app, or build your own psql binary from source, maybe even statically linking libpq etc, and then store that binary somewhere. If you just use it for some CI process, you don't need to update it all the time, and built once for macOS it will work for future versions as well.

If you do decide to go that route, `install_name_tool` and `otool -L` are going to help a lot. (read their man pages to understand what I mean)

**Comment by jakob** [comment]
Actually, just using brew or MacPorts is probably going to be the easiest solution. They should by default install pre-built binaries anyway.

**Comment by acecilia** [comment]
> They should by default install pre-built binaries anyway

I thought about it, but seems they are not prebuilt: brew is building it from a `zip` containing the sources. See the formula: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/libpq.rb#L4

**Comment by acecilia** [comment]
> We only provide the DMG files for download. It's a lot of effort to produce them for every release, since the process is only partially automated. Adding additional formats would be even more effort.

Is there any reason for using `dmg` and not `zip`?

**Comment by jakob** [comment]
> Is there any reason for using dmg and not zip?

Yes, because it must be moved to Applications folder before using. The dmg provides the link to /Applications, so people are less likely to just double click the app (which would fail).

> I thought about it, but seems they are not prebuilt

The formula tells brew how to build it from source, but you'll also see that there are hashes for pre-built binaries below it. (Homebrew calls pre-built binaries "bottles")

**Comment by acecilia** [comment]
@jakob The URLs for those prebuilt binaries I dont think are warranted to exist, and they can change

**Comment by jakob** [comment]
@acecilia Yes, you should just use homebrew to install psql and it will automatically download prebuilt binaries.

If you want to install pre-built binaries without using Homebrew or Macports, you need to come up with your own solution, and it's probably going to be a lot of work. Mounting a DMG vs extracting a ZIP is going to be the least of your problems

**Comment by acecilia** [comment]
Using homebrew is not an option for me because it installs globally and changes the state of the machine, and because it does not allow to pin which version of a dependency to install.

It is okey for users manually running the command, but not a good option for automated CI.

I can see that the app inside the DMG is not self contained and must be located exactly under `/Applications` for it to work 😞

**Comment by jakob** [comment]
@acecilia You can install them anywhere if you change the paths with `install_name_tool`.

**Comment by jakob** [comment]
See issue #338 for an example

**Comment by acecilia** [comment]
The problem with that is that absolute paths are being stored in the binary.

> build your own psql binary from source, maybe even statically linking libpq etc, and then store that binary somewhere

I think this may be the best approach. Do you have any guidance on how to do this?
